[FEATURE] Feature filter algorithm for processing models

m-kuhn opened the pull request:
## Description
Adds a new "feature filter" algorithm. This algorithm takes a list of expressions and
creates an output for each of them. Matching features from the input layer are sent
to the outputs.

This allows for an algorithm to only handle a subset of incoming features or to send
different features to different outputs.

The algorithm is only available in the modeler.

![image](https://user-images.githubusercontent.com/588407/38488524-0c138826-3be4-11e8-8022-bc2f3c8a6060.png)


## TODO

- [x] Make it work
- [x] Unit Tests
- [x] Allow outputs to be directly written to new layers
- [x] Have an expression builder widget to define the expression

m-kuhn committed:
[processing] Pass algorithm configuration on run

Because if it's not passed here, it's basically like writing it to /dev/null right from the beginning.

m-kuhn committed:
[processing] Allow algorithms to provide additional configuration widgets

m-kuhn committed:
[FEATURE] Add feature filter algorithm

Adds a new "feature filter" algorithm. This algorithm takes a list of expressions and
creates an output for each of them. Matching features from the input layer are sent
to the outputs.

This allows for an algorithm to only handle a subset of incoming features or to send
different features to different outputs.

The algorithm is only available in the modeler.

nyalldawson commented:
> Allow outputs to be directly written to new layers

What's this to-do mean?

> Unit tests

Heads up - some C++ native modeler algs have direct C++ unit tests. This is probably an easier approach to use here also and avoid the complexity of trying to fit the tests into the normal algorithm test framework

m-kuhn commented:
> >  Allow outputs to be directly written to new layers
> What's this to-do mean?

So far outputs can only be sent to other algorithms. If nothing is connected, anything sent there is gone for good.

> > Unit tests
> Heads up - some C++ native modeler algs have direct C++ unit tests. This is probably an easier approach to use here also and avoid the complexity of trying to fit the tests into the normal algorithm test framework

Is it not possible to call a model from the test framework? That would make it quite easy to test.

m-kuhn commented:
The hard part is the expression builder. Because currently the config widget is created within the analysis library, qgis_gui (and therefore the nice expression widgets) are not available. I am a bit worried that this will considerably complicate things.

nyalldawson commented:
> Is it not possible to call a model from the test framework? That would make it quite easy to test

You can, but adding specific granular tests would be good here (e.g. checking that outputs are correctly created after init with different configurations). That'd be a pain to do using standard algorithm tests.

nyalldawson commented on src/analysis/processing/qgsalgorithmfilter.cpp:
```diff
@@ -0,0 +1,247 @@
+
+QString QgsFilterAlgorithm::group() const
+{
+  return QObject::tr( "Vector geometry" );
```
This should be Vector table, not geometry

nyalldawson commented on src/analysis/processing/qgsalgorithmfilter.cpp:
```diff
@@ -0,0 +1,247 @@
+  if ( !source )
+    return QVariantMap();
+
+  QgsExpressionContext expressionContext = context.expressionContext();
```
Use createExpressionContext instead

nyalldawson commented on src/analysis/processing/qgsalgorithmfilter.cpp:
```diff
@@ -0,0 +1,247 @@
+  for ( const Output *output : qgis::as_const( mOutputs ) )
+  {
+    outputs.insert( output->name, output->destinationIdentifier );
+  }
```
I think you should be deleting the sinks here too

m-kuhn replied:
They are deleted in the destructor, but that can also be moved here.

nyalldawson replied:
I think it needs to be moved - some sinks do finalization code which probably should be done be within the processAlgorithm step

m-kuhn replied:
also done

m-kuhn committed:
Doxygen and license headers

m-kuhn committed:
A bit more RAII

m-kuhn committed:
Move processing algorithm configuration widget to gui library

This way we can also use all the fancy widgets like a QgsExpressionLineEdit.

m-kuhn committed:
Cleanup feature sinks directly after using them

m-kuhn committed:
Create proper expression context

m-kuhn committed:
Put feature filter algorithm in vector table group

m-kuhn committed:
Do not crash when using a parameter with FlagHidden

m-kuhn commented on src/core/processing/models/qgsprocessingmodelalgorithm.cpp:
```diff
@@ -89,9 +89,6 @@ QVariantMap QgsProcessingModelAlgorithm::parametersForChildAlgorithm( const QgsP
   Q_FOREACH ( const QgsProcessingParameterDefinition *def, child.algorithm()->parameterDefinitions() )
   {
-    if ( def->flags() & QgsProcessingParameterDefinition::FlagHidden )
-      continue;
```
@nyalldawson this fixes a crash for me as soon as I set the outputs to hidden (so I don't see them as parameters in the algorithm configuration widget). I am 0% sure this is the right approach to deal with this :D

nyalldawson commented:
Looking very nice!

The one thing I don't like here is that QgsProcessingAlgorithmConfigurationWidget and QgsProcessingAlgorithmConfigurationWidgetFactory are in core, when they really belong in gui.

What do you think about moving these to gui, and then creating a separate registry for the gui components of processing? That's how I handled this recently with the layouts engine - there's QgsLayoutItemRegistry in core (and an instance attached to QgsApplication) and QgsLayoutItemGuiRegistry in gui (with an instance attached to QgsGui).

This means there could be a total clean break between the processing gui and core components.

m-kuhn commented:
Sounds good to me, I was also thinking about that.

m-kuhn committed:
Move processing algorithm configuration code to gui lib

m-kuhn committed:
[processing] Show model outputs even if they are hidden in the child

m-kuhn committed:
API doc and typo fixes

m-kuhn committed:
Add missing @endcond

m-kuhn commented on src/core/processing/models/qgsprocessingmodelalgorithm.cpp:
```diff
@@ -769,6 +769,9 @@ void QgsProcessingModelAlgorithm::updateDestinationParameters()
       std::unique_ptr< QgsProcessingParameterDefinition > param( source->clone() );
+      // Even if an output was hidden in a child algorithm, we want to show it here for the final
+      // outputs.
+      param->setFlags( param->flags() & ~QgsProcessingParameterDefinition::FlagHidden );
```
Another very hacky-feeling thing to do. Does someone know what hidden parameters are for?

nyalldawson commented on src/analysis/processing/qgsalgorithmfilter.cpp:
```diff
@@ -92,6 +96,8 @@ QVariantMap QgsFilterAlgorithm::processAlgorithm( const QVariantMap &parameters,
   for ( Output *output : qgis::as_const( mOutputs ) )
   {
     output->sink.reset( parameterAsSink( parameters, output->name, context, output->destinationIdentifier, source->fields(), source->wkbType(), source->sourceCrs() ) );
+    if ( !output->sink )
```
I think this should throw a QgsProcessingException instead - it's a fatal error, we don't want to silently ignore it and continue with the model execution.

m-kuhn replied:
That was just copied from the buffer algorithm, I also found that a bit strange.

nyalldawson replied:
I'll fix that

m-kuhn replied:
Cool, thanks. The one for here is in the pipeline too

nyalldawson replied:
https://github.com/qgis/QGIS/commit/015665688262f95e148972d699fc2270d622a75c

nyalldawson commented:
:+1:

m-kuhn committed:
Add missing docstring

m-kuhn committed:
Throw exception when sink cannot be created

m-kuhn committed:
Fix use before assign

m-kuhn committed:
Make models available in processing tests

m-kuhn committed:
Add filter feature processing test

m-kuhn committed:
for : in

m-kuhn committed:
fixup testdata

m-kuhn committed:
Fix api doc and sip issues

m-kuhn committed:
Plug memory leak

m-kuhn committed:
Show expression builder and checkbox in filter configuration

m-kuhn committed:
Rename "Filter" to "Feature Filter"

m-kuhn committed:
Better error reporting

m-kuhn committed:
Do not complain on invalid geometries

m-kuhn commented on src/analysis/processing/qgsalgorithmfilter.cpp:
```diff
@@ -88,6 +88,7 @@ void QgsFilterAlgorithm::initAlgorithm( const QVariantMap &configuration )
 
 QVariantMap QgsFilterAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
 {
+  context.setInvalidGeometryCheck( QgsFeatureRequest::GeometryNoCheck );
```
@nyalldawson is this the correct place to disable this check? I don't think it's used anywhere so far...

m-kuhn committed:
Properly initialize filter configuration table

m-kuhn committed:
Immediately apply configuration after change

m-kuhn committed:
Really do not care about invalid geometries

in feature filter algorithm. We just leave the geometry alone in whatever
state it may be. If the consumers of our features have a problem with that,
let them take care of it. Just because THEY do not like invalid geometries
does not mean WE have to adjust our behavior here.

m-kuhn committed:
Add more feature filter algorithm tests

m-kuhn committed:
Add a remaining docstring

m-kuhn commented:
Any final comments on this?

nyalldawson commented:
Looks fantastic to me! Could you add tests covering QgsProcessingGuiRegistry and then I'm a big +1 to merge.

m-kuhn committed:
Protect the const-ituional correctness

m-kuhn committed:
Add processing gui tests
